cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/spec/PipelineSpecGenerator.java

```diff
@@ -296,28 +281,41 @@ private ConfiguredStage configureStage(ETLStage stage, ValidatedPipeline validat
         singlePlugin.configurePipeline(pipelineConfigurer);
       }
     } catch (InvalidConfigPropertyException e) {
-      throw new InvalidPipelineException(new InvalidConfigPropertyError(stageName, e), e);
+      stageConfigurer.addFailure(e.getMessage(),
+                                 String.format("Provide valid value for config property '%s'.", e.getProperty()))
+        .withInvalidStageConfigCause(e.getProperty());
     } catch (InvalidStageException e) {
       if (e.getReasons().isEmpty()) {
-        throw new InvalidPipelineException(new StageValidationError(e.getMessage(), stageName), e);
+        stageConfigurer.addFailure(e.getMessage(), null);
       }
-      List<StageValidationError> errors = new ArrayList<>(e.getReasons().size());
+
       for (InvalidStageException reason : e.getReasons()) {
         if (reason instanceof InvalidConfigPropertyException) {
-          errors.add(new InvalidConfigPropertyError(stageName, (InvalidConfigPropertyException) reason));
+          InvalidConfigPropertyException configException = (InvalidConfigPropertyException) reason;
+          stageConfigurer.addFailure(configException.getMessage(),
+                                     String.format("Provide valid value for config property '%s'.",
+                                                   configException.getProperty()))
+            .withInvalidStageConfigCause(configException.getProperty());
         } else {
-          errors.add(new StageValidationError(reason.getMessage(), stageName));
+          stageConfigurer.addFailure(reason.getMessage(), null);
         }
       }
-      throw new InvalidPipelineException(errors, e);
+    } catch (ValidationException e) {
+      throw e;
+    } catch (NullPointerException e) {
+      // handle the case where plugin throws null pointer exception, this is to avoid having 'null' as error message
+      stageConfigurer.addFailure(getNpeErrorMessage(stageName, e),
+                                 "Make sure optional configurations are handled by the plugin implementation.");
```

Comment:
NPE usually isn't something the pipeline developer can fix, but is an indication that plugin code has a bug. It's not always something about optional configurations either.

Changing it to a validation failure also means the stack trace will get hidden, which makes it more difficult for the plugin developer to track down the actual bug. Maybe it makes sense to have some sort of null pointer cause that has the stack trace as an attribute?

Reply:
Another approach is we can let NPE get propagated with modified message. This is because capturing it as a failure would not help pipeline user fix the issue. However, plugin developer can still see the stacktrace in the logs in order to fix the underlying issue with the plugin. Having stacktrace as a cause may not be helpful for plugin developer since it will not be displayed by UI on the stage as error.

Reply:
letting it propagate means the service call will return a 500 error and the pipeline developer won't get any feedback about what is wrong. Exceptions thrown by plugin code should always be handled by the app somehow. Only exceptions that come from the app itself should result in a 500 error.

Whether or not the stacktrace gets shown is up to the UI, but it's a useful piece of information that shouldn't get dropped. The corrective action also shouldn't mention optional configurations. It can probably be null for now, not sure anything useful can be put there.